### PR TITLE
feat: Refine Testing Configuration

### PR DESCRIPTION
**Pull request by williamputraintan**
Put back GH Action which runs on AWS CodeBuild as GH Runners except for lint and Security check. All application test runs at different CodeBuild runner to cut on time.

### CODE REVIEW

**Comment by williamputraintan** [comment]
Well, I guess we really need to setup self hosted runners in CodeBuild. GH Runners do not support arm architecture 😐

**Comment by victorskl** [comment]
Go for it, Will. CodeBuild runner sounds good. 

At least we will learn a thing or two for trying it. ACK; it is a new feature between GitHub Action and AWS CodeBuild side of eco-system. 

Plus, if it works well, no worries about potential split billing issue about compute cost on building the software stacks. Compute cost will be consolidated at AWS end.

We achieve best of both worlds then.


FYI @ohofmann @reisingerf

**Comment by victorskl** [comment]
Related to #100 story from the Platform engineering backlog

File: lib/pipeline/statelessPipelineStack.ts [hunk 1]
```diff
@@ -20,6 +20,41 @@ export class StatelessPipelineStack extends cdk.Stack {
...
+        image: 'aws/codebuild/amazonlinux2-aarch64-standard:3.0',
+        privilegedMode: true,
+      },
+      source: {
+        type: 'GITHUB',
+        gitCloneDepth: 1,
+        location: 'https://github.com/umccr/orcabus.git',
+        reportBuildStatus: false,
+      },
+      logsConfig: { cloudWatchLogs: { status: 'DISABLED' } },
```
Comment (williamputraintan) [inline lib/pipeline/statelessPipelineStack.ts:50]: Not seeing anything is useful here, might as well disabled this.

Screenshot 2024-05-13 at 17 42 27

File: lib/pipeline/statelessPipelineStack.ts [hunk 2]
```diff
@@ -20,6 +20,41 @@ export class StatelessPipelineStack extends cdk.Stack {
   constructor(scope: Construct, id: string, props: cdk.StackProps) {
     super(scope, id, props);
 
+    // Define CodeBuild project for GH action runner to use
+    // the GH repo defined below already configured to allow CB webhook
+    // This is actually not part of the pipeline, so I guess we could move this someday.
+    const ghRunnerRole = new iam.Role(this, 'GHRunnerRole', {
+      assumedBy: new iam.ServicePrincipal('codebuild.amazonaws.com'),
+    });
+    new codebuild.CfnProject(this, 'GHRunnerCodeBuildProject', {
```
Comment (williamputraintan) [inline lib/pipeline/statelessPipelineStack.ts:29]: Well, this webhook GH runner <> AWS CodeBuild is [new](https://aws.amazon.com/about-aws/whats-new/2024/04/aws-codebuild-managed-github-action-runners/), so is not a surprised that the webhook flag for L2 CDK construct hasn't exist yet. Defining here the L1 construct.

[cdk-issue-#30062](https://www.github.com/aws/aws-cdk/issues/30062)

File: lib/pipeline/statelessPipelineStack.ts [hunk 3]
```diff
@@ -20,6 +20,41 @@ export class StatelessPipelineStack extends cdk.Stack {
...
+      triggers: {
+        webhook: true,
+        buildType: 'BUILD',
+        filterGroups: [[{ type: 'EVENT', pattern: 'WORKFLOW_JOB_QUEUED' }]],
+      },
+    });
+
     // A connection where the pipeline get its source code
     const codeStarArn = ssm.StringParameter.valueForStringParameter(this, 'codestar_github_arn');
     const sourceFile = pipelines.CodePipelineSource.connection('umccr/orcabus', 'main', {
```
Comment (williamputraintan) [inline lib/pipeline/statelessPipelineStack.ts:60]: On the unittest below, did want to try the [batch/build-list](https://docs.aws.amazon.com/codebuild/latest/userguide/batch-build-buildspec.html#build-spec.batch.build-list), so could speed up by parallel testting, but might not be convenient yet for cdk-pipeline codebuild source. a feature request was put to cdk issue [#29592](https://www.github.com/aws/aws-cdk/issues/29592), smth that we could look for in the future.

File: lib/pipeline/statelessPipelineStack.ts [hunk 4]
```diff
@@ -20,6 +20,41 @@ export class StatelessPipelineStack extends cdk.Stack {
...
 
+    // Define CodeBuild project for GH action runner to use
+    // the GH repo defined below already configured to allow CB webhook
+    // This is actually not part of the pipeline, so I guess we could move this someday.
+    const ghRunnerRole = new iam.Role(this, 'GHRunnerRole', {
+      assumedBy: new iam.ServicePrincipal('codebuild.amazonaws.com'),
+    });
+    new codebuild.CfnProject(this, 'GHRunnerCodeBuildProject', {
+      // the name here act as a unique id for GH action to know which CodeBuild to use
+      // So if you change this, you need to update the GH action yaml file (.github/workflows/prbuild.yml)
```
Comment (reisingerf) [inline lib/pipeline/statelessPipelineStack.ts:31]: Huh? 
The name here does not match the name in the prbuild.yml file though. How are they matched up?

File: lib/pipeline/statelessPipelineStack.ts (hunk 1 above)
Comment (reisingerf) [reply lib/pipeline/statelessPipelineStack.ts:50]: Hm, the useful bits may only come when something goes wrong? 
Are those logs visible via the codebuild history? 
In any case, I guess easy to add back in if we need to...

**Review by reisingerf — APPROVED** [review]
Let's try!

File: lib/pipeline/statelessPipelineStack.ts (hunk 1 above)
Comment (williamputraintan) [reply lib/pipeline/statelessPipelineStack.ts:50]: The real testing logs goes to the GH action itself.

https://github.com/umccr/orcabus/actions/runs/9063282042

File: lib/pipeline/statelessPipelineStack.ts (hunk 4 above)
Comment (williamputraintan) [reply lib/pipeline/statelessPipelineStack.ts:31]: Ahh so the pattern on the GH action side is below, so the project name should be the exact name. I'll update this on the commenting

```
runs-on: codebuild-<project-name>-${{ github.run_id }}-${{ github.run_attempt }}
```

File: lib/pipeline/statelessPipelineStack.ts (hunk 1 above)
Comment (williamputraintan) [reply lib/pipeline/statelessPipelineStack.ts:50]: To enable just need to change the `DISABLED` to `ENABLED`, and need to add log group (and its respective role) as this is the L1 construct

**Review by victorskl — APPROVED** [review]
Good job, Will.

**Comment by williamputraintan** [comment]
cc @alexiswl @brainstorm @mmalenic @raylrui 

Just a side note if you are not aware of the `[skip ci]` flag (when added part of your commit message), it lets you skip the GH Action Jobs. Useful if you know your commit do not break any tests (e.g. code commenting commit). Eventually before it gets deployed, the code will get tested again.

https://docs.github.com/en/actions/managing-workflow-runs/skipping-workflow-runs

**Pull request merged**